Source/Core/Common/MemoryUtil.cpp

```diff
@@ -29,46 +29,113 @@
+      last_executable_addr = reinterpret_cast<uintptr_t>(info.AllocationBase) - size;
+      continue;
+    }
+
+    const auto base_address_value = reinterpret_cast<uintptr_t>(info.BaseAddress);
+    if (base_address_value + info.RegionSize >= end && base_address_value <= last_executable_addr)
+      return reinterpret_cast<void*>(last_executable_addr);
+
+    last_executable_addr -= size;
+  }
+
+  return nullptr;
+}
+#endif
+
 // This is purposely not a full wrapper for virtualalloc/mmap, but it
 // provides exactly the primitive operations that Dolphin needs.
 
-void* AllocateExecutableMemory(size_t size, bool low)
+void* AllocateExecutableMemory(size_t size)
 {
 #if defined(_WIN32)
-  void* ptr = VirtualAlloc(0, size, MEM_COMMIT, PAGE_EXECUTE_READWRITE);
-#else
-  static char* map_hint = nullptr;
-#if defined(_M_X86_64) && !defined(MAP_32BIT)
-  // This OS has no flag to enforce allocation below the 4 GB boundary,
-  // but if we hint that we want a low address it is very likely we will
-  // get one.
-  // An older version of this code used MAP_FIXED, but that has the side
-  // effect of discarding already mapped pages that happen to be in the
-  // requested virtual memory range (such as the emulated RAM, sometimes).
-  if (low && (!map_hint))
-    map_hint = (char*)RoundPage(512 * 1024 * 1024); /* 0.5 GB rounded up to the next page */
+  void* ptr;
+#if defined(_M_X86_64)
+  if (sys_info.dwPageSize == 0)
+    GetSystemInfo(&sys_info);
+
+  if (hint_location > 0xFFFFFFFFULL)
+  {
+    size_t aligned_size = RoundPage(size);
+    ptr = SearchForFreeMem(aligned_size);
+    if (!ptr)
+    {
+      // Let's try again, from the top.
+      // When we deallocate, this doesn't change, so we eventually run out of space.
+      last_executable_addr = 0;
+      ptr = SearchForFreeMem(aligned_size);
+    }
+    if (ptr)
+      ptr = VirtualAlloc(ptr, aligned_size, MEM_RESERVE | MEM_COMMIT, PAGE_EXECUTE_READWRITE);
+    else
+      ERROR_LOG(MEMMAP, "Unable to find nearby executable memory for jit");
+  }
+  else
 #endif
-  void* ptr = mmap(map_hint, size, PROT_READ | PROT_WRITE | PROT_EXEC, MAP_ANON | MAP_PRIVATE
-#if defined(_M_X86_64) && defined(MAP_32BIT)
-                                                                           | (low ? MAP_32BIT : 0)
+  {
+    ptr = VirtualAlloc(0, size, MEM_RESERVE | MEM_COMMIT, PAGE_EXECUTE_READWRITE);
+  }
+#else
+  static uintptr_t map_hint;
+#if defined(_M_X86_64)
+  // Try to request one that is close to our memory location if we're in high memory.
+  // We use a dummy global variable to give us a good location to start from.
+  if (!map_hint)
+  {
+    if (hint_location > 0xFFFFFFFFULL)
+      map_hint =
+          RoundPage(hint_location) - 0x20000000;  // 0.5gb lower than our approximate location
+    else
+      map_hint = 0x20000000;  // 0.5GB mark in memory
+  }
+  else if (map_hint > 0xFFFFFFFFULL)
+  {
+    map_hint -= RoundPage(size); /* round down to the next page if we're in high memory */
```

Comment:
This looks like it has been moved from elsewhere and extended, but we generally use `//` comments over `/* */` blocks. Could you please change that while you're at it?